src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs

```diff
@@ -1122,9 +1172,77 @@ private static ImmutableArray<StateForCase> RemoveEvaluation(ImmutableArray<Stat
             trueTestImpliesTrueOther = false;
             falseTestImpliesTrueOther = false;
 
+            trueOther = Tests.True.Instance;
+
             // if the tests are for unrelated things, there is no implication from one to the other
             if (!test.Input.Equals(other.Input))
-                return;
+            {
+                switch (test.Input.Source, other.Input.Source)
+                {
+                    case (BoundDagPropertyEvaluation { IsLengthOrCount: true } s1, BoundDagPropertyEvaluation { IsLengthOrCount: true } s2):
+                        {
+                            (BoundDagTemp s1Input, _, int s1Offset) = GetCanonicalInput(s1);
+                            (BoundDagTemp s2Input, _, int s2Offset) = GetCanonicalInput(s2);
+                            if (s1Input.Equals(s2Input))
+                            {
+                                // This happens when we have a length test in a slice pattern, in which case we align
+                                // the values to the current offset and continue. This is because we will update the
+                                // outer list length values based on a nested test but that hasn't happened yet.
+                                Debug.Assert(whenTrueValues is not null);
+                                Debug.Assert(whenFalseValues is not null);
+                                whenTrueValues = ValueSetFactory.Shift(whenTrueValues, s1Offset - s2Offset);
+                                whenFalseValues = whenTrueValues.Complement();
+                                break;
+                            }
+                        }
+                        return;
+                    case (BoundDagIndexerEvaluation s1, BoundDagIndexerEvaluation s2):
+                        {
+                            (BoundDagTemp s1Input, BoundDagTemp s1LengthTemp, int s1Index) = GetCanonicalInput(s1);
+                            (BoundDagTemp s2Input, BoundDagTemp s2LengthTemp, int s2Index) = GetCanonicalInput(s2);
+                            Debug.Assert(s1LengthTemp.Syntax is ListPatternSyntax);
+                            Debug.Assert(s2LengthTemp.Syntax is ListPatternSyntax);
+                            if (s1Input.Equals(s2Input) &&
+                                // We don't want to pair two indices within the same pattern.
+                                s1LengthTemp.Syntax != s2LengthTemp.Syntax)
```

Comment:
:memo: This is needed for something like `[1, ..[2, ..[], 3]]`, `[0]` and `[^1]` *could* subsume each other but not within the same pattern. Note that for `[1, .., 3]` we correctly bail because length requirements are not met.